src/Datasource/EntityTrait.php

```diff
@@ -197,14 +197,48 @@ public function __unset(string $field): void
      * $entity->set('name', 'Andrew');
      * ```
      *
-     * It is also possible to mass-assign multiple fields to this entity
-     * with one call by passing a hashed array as fields in the form of
-     * field => value pairs
+     * Some times it is handy to bypass setter functions in this entity when assigning
+     * fields. You can achieve this by disabling the `setter` option using the
+     * `$options` parameter:
+     *
+     * ```
+     * $entity->set('name', 'Andrew', ['setter' => false]);
+     * ```
+     *
+     * You can use the `asOriginal` option to set the given field as original, if it wasn't
+     * present when the entity was instantiated.
+     *
+     * ```
+     * $entity = new Entity(['name' => 'andrew', 'id' => 1]);
+     *
+     * $entity->set('phone_number', '555-0134');
+     * print_r($entity->getOriginalFields()) // prints ['name', 'id']
+     *
+     * $entity->set('phone_number', '555-0134', ['asOriginal' => true]);
+     * print_r($entity->getOriginalFields()) // prints ['name', 'id', 'phone_number']
+     * ```
+     *
+     * @param string $field The name of field to set.
+     * @param mixed $value The value to set to the field.
+     * @param array<string, mixed> $options Options to be used for setting the field. Allowed option
+     * keys are `setter`, `guard` and `asOriginal`
+     * @return $this
+     * @throws \InvalidArgumentException
+     */
+    public function set(string $field, mixed $value, array $options = []): static
+    {
+        $options += ['guard' => false];
+
+        return $this->patch([$field => $value], $options);
+    }
```

Comment:
We'll need a deprecation in 5.x for the change in behavior. We could also introduce `patch()` in 5.x so that upgrading is easier for applications.

I can take care of the backport and deprecation if you'd like.

Reply:
I can update `5.next` first, but I am unsure if we should be triggering explicit warning for using `set()` with an array.

We use `EntityInterface` to type arguments/returns and the interface doesn't have the `patch()` method. We can't assume that all implementations of the `EntityInterface` are done using the `EntityTrait`.

Reply:
> I can update 5.next first, but I am unsure if we should be triggering explicit warning for using set() with an array.

We could use `method_exists()` in each place we currently call `set()` but that could get messy.

> We use EntityInterface to type arguments/returns and the interface doesn't have the patch() method.

Couldn't a `@method` annotation help here at least for the typecheckers.

> We can't assume that all implementations of the EntityInterface are done using the EntityTrait.

Yeah, we could use `method_exists()`, but it could get jank. We could also add patch() in 5.x, and only apply an annotation deprecation, and then make it a formal deprecation in 6.x when the interface can be updated :shrug:

Reply:
Having to add `method_exists()` check everywhere was my main gripe, certainly don't want to do that. Can't use the `@deprecated` annotation either since it can't be used to denote the deprecation of a particular type for an argument.

Reply:
Another solution would be to only add `patch()` in 6.x where we can update the interface (Like you have currently). We also preserve compatibility of behavior for `set()` however, `EntityTrait` begins emitting a deprecation when `$field` is an array. We can also provide a rector rule that rewrites any `set()` calls with array literals to be `patch()` to reduce the upgrade noise.

This approach has the drawback of requiring another future major release to clean up `set()`, but it would let folks upgrade to 6.x without requiring a significant rewrite. Deprecations can be disabled on a path by path basis already so incremental progress can be made as folks have time to incrementally upgrade.

Reply:
I am now leaning towards wrapping the `set()` calls with array in `method_exists()` checks rather than waiting 2 major release to get this done :slightly_smiling_face:.

Reply:
There aren't a *ton* of calls to `set()` in the framework at least. Plugins will likely have some more though.